File: java edition/topSort.java
import java.util.*;
public class topSort{
    public static void main(String[] args){
        ArrayList<DirectedGraphNode> test = new ArrayList<DirectedGraphNode>();
        // nodes
        DirectedGraphNode a = new DirectedGraphNode(0);
        DirectedGraphNode b = new DirectedGraphNode(1);
        DirectedGraphNode c = new DirectedGraphNode(2);
        DirectedGraphNode d = new DirectedGraphNode(3);
        DirectedGraphNode e = new DirectedGraphNode(4);
        DirectedGraphNode f = new DirectedGraphNode(5);

        // edges
        a.neighbors.add(b);
        a.neighbors.add(c);
        a.neighbors.add(d);

        b.neighbors.add(e);

        c.neighbors.add(e);
        c.neighbors.add(f);

        d.neighbors.add(e);
        d.neighbors.add(f);

        test.add(a); test.add(b);
        test.add(c); test.add(b);
        test.add(e); test.add(f);

        ArrayList<DirectedGraphNode> testRes = topSort(test);
        for(DirectedGraphNode i : testRes){
            System.out.println(i.label);
        }
    }

    public static ArrayList<DirectedGraphNode> topSort(ArrayList<DirectedGraphNode> graph){
        ArrayList<DirectedGraphNode> result = new ArrayList<DirectedGraphNode>();
        if(graph == null || graph.size() == 0){
            return result;
        }
        HashMap<DirectedGraphNode, Integer> hm = new HashMap<DirectedGraphNode, Integer>();
        for(DirectedGraphNode node : graph){
            for(DirectedGraphNode neighbor : node.neighbors){
                if(!hm.containsKey(neighbor)){
                    hm.put(neighbor, 1);
                }else{
                    hm.put(neighbor, hm.get(neighbor) + 1);
                }
            }
        }

        Queue<DirectedGraphNode> q = new LinkedList<DirectedGraphNode>();
        for(DirectedGraphNode node : graph){
            if(!hm.containsKey(node)){
                q.offer(node);
                result.add(node);
            }
        }

        while(!q.isEmpty()){
            DirectedGraphNode cur = q.poll();
            for(DirectedGraphNode neighbor : cur.neighbors){
                hm.put(neighbor, hm.get(neighbor) - 1);
                if(hm.get(neighbor) == 0){
                    q.offer(neighbor);
                    result.add(neighbor);
                }
            }
        }
        return result;
    }
}

File: topSort.py
from sets import Set

# Definition for a Directed graph node
class DirectedGraphNode:
   def __init__(self, x):
       self.label = x
       self.neighbors = []

class Queue:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def enqueue(self,item):
        self.items.insert(0,item)

    def dequeue(self):
        return self.items.pop()

    def size(self):
        return len(self.items)

class Solution:
    def topSort(self, graph):
        res = []
        if not graph:
            return res

        allNodes = Set(graph)
        nbs = Set([])

        for n in graph:
            for nb in n.neighbors:
                nbs.add(nb)

        start = allNodes - nbs
        if len(start) == 1:
            startNode = list(start)[0]

        # find start node and do bfs
        q = Queue()
        visited = Set([])

        q.enqueue(startNode)
        while not q.isEmpty():
            node = q.dequeue()
            visited.add(node)
            for i in node.neighbors:
                if i in visited:
                    continue
                else:
                    q.enqueue(i)

        res = list(visited)

        return res

a = Solution()

a0 = DirectedGraphNode(0)
a1 = DirectedGraphNode(1)
a2 = DirectedGraphNode(2)
a3 = DirectedGraphNode(3)
a4 = DirectedGraphNode(4)
a5 = DirectedGraphNode(5)

a0.neighbors += [a1, a2, a3]
a1.neighbors.append(a4)
a2.neighbors += [a4, a5]
a3.neighbors += [a4, a5]

test = []
test += [a0, a1, a2, a3, a4, a5]
result = a.topSort(test)
for i in result:
    print i.label
